File: backend/internal/crypto/canonical_test.go
package crypto

import (
	"bytes"
	"testing"
	"time"
)

func TestCanonicalizeSortsKeys(t *testing.T) {
	v := map[string]any{"b": 2, "a": 1, "c": 3}
	got, err := Canonicalize(v)
	if err != nil {
		t.Fatal(err)
	}
	want := []byte(`{"a":1,"b":2,"c":3}`)
	if !bytes.Equal(got, want) {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestCanonicalizeNested(t *testing.T) {
	type inner struct {
		Z int `json:"z"`
		A int `json:"a"`
	}
	type outer struct {
		M inner `json:"m"`
		K int   `json:"k"`
	}
	got, err := Canonicalize(outer{M: inner{Z: 1, A: 2}, K: 3})
	if err != nil {
		t.Fatal(err)
	}
	want := []byte(`{"k":3,"m":{"a":2,"z":1}}`)
	if !bytes.Equal(got, want) {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestCanonicalizeDeterministic(t *testing.T) {
	type payload struct {
		ID     string    `json:"id"`
		When   time.Time `json:"when"`
		Amount int64     `json:"amount"`
		Bytes  []byte    `json:"bytes"`
	}
	p := payload{
		ID:     "user_1",
		When:   time.Date(2026, 4, 13, 12, 0, 0, 0, time.UTC),
		Amount: 50000,
		Bytes:  []byte{0x01, 0x02, 0x03},
	}
	a, err := Canonicalize(p)
	if err != nil {
		t.Fatal(err)
	}
	for range 50 {
		b, err := Canonicalize(p)
		if err != nil {
			t.Fatal(err)
		}
		if !bytes.Equal(a, b) {
			t.Fatalf("non-deterministic: %s vs %s", a, b)
		}
	}
}

func TestCanonicalizeBytesBase64(t *testing.T) {
	v := map[string]any{"k": []byte{0xff, 0x00, 0x7f}}
	got, err := Canonicalize(v)
	if err != nil {
		t.Fatal(err)
	}
	// json.Marshal encodes []byte as standard base64.
	want := []byte(`{"k":"/wB/"}`)
	if !bytes.Equal(got, want) {
		t.Errorf("got %s, want %s", got, want)
	}
}

File: mobile/packages/core/test/canonical_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:offlinepay_core/src/canonical.dart';
import 'package:test/test.dart';

void main() {
  test('sorts object keys lexicographically', () {
    final got = canonicalize({'b': 2, 'a': 1, 'c': 3});
    expect(utf8.decode(got), equals('{"a":1,"b":2,"c":3}'));
  });

  test('sorts nested object keys', () {
    final got = canonicalize({
      'm': {'z': 1, 'a': 2},
      'k': 3,
    });
    expect(utf8.decode(got), equals('{"k":3,"m":{"a":2,"z":1}}'));
  });

  test('deterministic across runs', () {
    final value = {
      'id': 'user_1',
      'when': DateTime.utc(2026, 4, 13, 12, 0, 0),
      'amount': 50000,
      'bytes': Uint8List.fromList([0x01, 0x02, 0x03]),
    };
    final a = canonicalize(value);
    for (var i = 0; i < 50; i++) {
      final b = canonicalize(value);
      expect(b, equals(a));
    }
  });

  test('byte arrays as standard base64', () {
    final got = canonicalize({
      'k': const CanonicalBytes([0xff, 0x00, 0x7f]),
    });
    expect(utf8.decode(got), equals('{"k":"/wB/"}'));
  });

  test('RFC3339Nano formatting trims trailing zeros', () {
    expect(
      formatRfc3339Nano(DateTime.utc(2026, 4, 13, 12, 0, 0)),
      equals('2026-04-13T12:00:00Z'),
    );
    expect(
      formatRfc3339Nano(DateTime.utc(2026, 4, 13, 12, 0, 0, 500)),
      equals('2026-04-13T12:00:00.5Z'),
    );
  });
}
